### PR TITLE
20230517-linuxkm-benchmarks

### PR DESCRIPTION
**Pull request by douzzer**
refactor `benchmark.c` for linux kernel compatibility -- `WOLFSSL_SMALL_STACK` and `WOLFSSL_NO_FLOAT_FMT` codepaths, `SAVE`/`RESTORE_VECTOR_REGISTERS`, refactor of several stack array initializations that broke in the kernel, and replacement of an `fputs()` call with `printf()`;

`configure.ac`: add `--enable-linuxkm-benchmarks`; add check for `async.{c,h}` when `--enable-asynccrypt`; update failure message for the opensslextra `AC_CHECK_HEADER()` test;

`linuxkm/linuxkm_memory.c`: refactor `SAVE`/`RESTORE_VECTOR_REGISTERS()` to be per-process rather than per-CPU, and add `migrate_disable`/`enable()` to `kernel_fpu_begin`/`end()` because `preempt_disable()` is just a barrier on `_PREEMPT_VOLUNTARY` kernels;

`linuxkm/linuxkm_wc_port.h`: activate `SAVE`/`RESTORE_VECTOR_REGISTERS()` whenever `defined(WOLFSSL_LINUXKM_USE_SAVE_VECTOR_REGISTERS)` for `benchmark.c` support, independent of vector crypto features;

`linuxkm/linuxkm_wc_port.h`: fix and optimize various alignment issues with stack and heap allocations;

fix macro definitions for `XMALLOC`/`XREALLOC`/`XFREE` to correctly use `kvmalloc` and friends when `defined(HAVE_KVMALLOC)`, and to use `wolfSSL_Malloc()` and friends when `defined(WOLFSSL_TRACK_MEMORY)`;

purge stale `LINUXKM_SIMD_IRQ` code;

`wolfssl/wolfcrypt/mem_track.h`: refactor for linuxkm compatibility, mainly by supporting `NO_STDIO_FILESYSTEM`;

`wolfcrypt/src/aes.c`: in `wc_AesSetKeyLocal()`, add an alignment check in the `haveAESNI` path for `WOLFSSL_LINUXKM`, because the failure mode is module crash.

also various unrelated point fixes for analyzer carps, one commit per fix.

tested with `wolfssl-multi-test.sh ... super-quick-check '.*linuxkm.*' benchmark-wolfcrypt-intelasm-all benchmark-wolfcrypt-noasm-all`

### CODE REVIEW

File: wolfcrypt/benchmark/benchmark.c [hunk 1]
```diff
@@ -7468,66 +7629,91 @@ void bench_ecc(int useDeviceID, int curveId)
 #ifdef HAVE_ECC_ENCRYPT
 void bench_eccEncrypt(int curveId)
 {
-    ecc_key userA, userB;
+#define BENCH_ECCENCRYPT_MSG_SIZE 48
+#define BENCH_ECCENCRYPT_OUT_SIZE (BENCH_ECCENCRYPT_MSG_SIZE + WC_SHA256_DIGEST_SIZE + (MAX_ECC_BITS+3)/4 + 2)
```
Comment (dgarske) [inline wolfcrypt/benchmark/benchmark.c:7633]: Line length

**Review by dgarske — CHANGES_REQUESTED** [review]

File: linuxkm/module_hooks.c [hunk 2]
```diff
@@ -113,6 +113,15 @@ static void lkmFipsCb(int ok, int err, const char* hash)
 static int updateFipsHash(void);
 #endif
 
+#ifdef WOLFSSL_LINUXKM_BENCHMARKS
+#undef HAVE_PTHREAD
+#define STRING_USER
+#define NO_MAIN_FUNCTION
+#define current_time benchmark_current_time
+#define WOLFSSL_NO_FLOAT_FMT
+#include "/home/douzzer/com/wolfssl/src/wolfssl/wolfcrypt/benchmark/benchmark.c"
```
[the remark this answers is not in the document]
Comment (douzzer) [reply linuxkm/module_hooks.c:122]: LOL!!!

thank you @dgarske :grin:

File: wolfcrypt/benchmark/benchmark.c [hunk 3]
```diff
@@ -65,6 +65,28 @@
 #include <wolfssl/wolfcrypt/ecc.h>
 #include <wolfssl/wolfcrypt/wolfmath.h>
 
+#ifdef WOLFSSL_NO_FLOAT_FMT
+    #define FLT_FMT "%0ld,%09lu"
+    #define FLT_FMT_PREC "%0ld.%0*lu"
+    #define FLT_FMT_PREC2 FLT_FMT_PREC
+    #define FLT_FMT_ARGS(x) (long)(x), ((x) < 0) ? (unsigned long)(-(((x) - (double)(long)(x)) * 1000000000.0)) : (unsigned long)(((x) - (double)(long)(x)) * 1000000000.0)
```
[the remark this answers is not in the document]
Comment (douzzer) [reply wolfcrypt/benchmark/benchmark.c:72]: yeah sure I'll fix that -- but note there's a ton of overlong line in benchmark.c.  it's a bit of a shambles...

File: wolfcrypt/benchmark/benchmark.c [hunk 4]
```diff
@@ -6070,7 +6104,7 @@ void bench_hmac_md5(int useDeviceID)
 
 void bench_hmac_sha(int useDeviceID)
 {
-    byte key[] = { 0x0b, 0x0b, 0x0b, 0x0b, 0x0b, 0x0b, 0x0b, 0x0b,
+    static const byte key[] = { 0x0b, 0x0b, 0x0b, 0x0b, 0x0b, 0x0b, 0x0b, 0x0b,
```
[the remark this answers is not in the document]
Comment (douzzer) [reply wolfcrypt/benchmark/benchmark.c:6107]: crashes without `static`.  the problem is stack initialization.  exact root cause unknown, but suspect alignment weirdness.

File: wolfcrypt/src/aes.c [hunk 5]
```diff
@@ -3129,6 +3129,12 @@ static WARN_UNUSED_RESULT int wc_AesDecrypt(
             checkAESNI = 1;
         }
         if (haveAESNI) {
+            #ifdef WOLFSSL_LINUXKM
+            /* runtime alignment check */
+            if ((unsigned long)&aes->key & 0xf) {
```
[the remark this answers is not in the document]
Comment (douzzer) [reply wolfcrypt/src/aes.c:3134]: Only the bottom 4 bits matter here, so I figured casting to `unsigned long` was fine and keeps it maximally efficient.

But we do actually have a right way to do this -- `wc_ptr_t`, set up in types.h.  Will use that.

Btw, `size_t` is not reliably the same size as a pointer -- ran into this recently in-real-life.  That can produce inefficiencies or truncation warnings depending on which way it goes.

**Comment by douzzer** [comment]
Jenkins, retest this please.

**Review by dgarske — APPROVED** [review]

**Comment by douzzer** [comment]
note for posterity: the CI failure on the final commit in this PR is a Jenkins/github state model conflict.  the single subtest at issue, `PRB-fips-repo-and-harness-test-v2`, was hit by a github network glitch on 1st try, and succeeded (https://cloud.wolfssl-test.com/jenkins/job/PRB-fips-repo-and-harness-test-v2/5248/) on 2nd try.

**Pull request merged**